File: mirai-core/src/main/java/net/mamoe/mirai/utils/setting/MiraiSettingListSection.java
package net.mamoe.mirai.utils.setting;


import org.ini4j.Profile;

import java.io.IOException;
import java.util.Vector;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReentrantLock;


/**
 * @author NaturalHG
 */
public class MiraiSettingListSection extends Vector<Object> implements MiraiSettingSection {
    private Lock lock = new ReentrantLock();

    @SuppressWarnings("unchecked")
    public <T> T getAs(int index){
        return (T)super.get(index);
    }

    public int getInt(int index){
        return this.getAs(index);
    }

    public int getDouble(int index){
        return this.getAs(index);
    }

    public int getString(int index){
        return this.getAs(index);
    }

    public int getFloat(int index) {
        return this.getAs(index);
    }

    @Override
    public synchronized void saveAsSection(Profile.Section section) {
        section.clear();
        AtomicInteger integer = new AtomicInteger(0);
        this.forEach(a -> {
            section.put(String.valueOf(integer.getAndAdd(1)),a);
        });
    }

    @Override
    public void close() throws IOException {

    }

}

File: mirai-core/src/jvmMain/kotlin/net/mamoe/mirai/utils/setting/MiraiSettingListSection.kt
package net.mamoe.mirai.utils.setting


import org.ini4j.Profile
import java.io.IOException
import java.util.*
import java.util.concurrent.atomic.AtomicInteger
import java.util.concurrent.locks.ReentrantLock


/**
 * @author NaturalHG
 */
class MiraiSettingListSection : Vector<Any>(), MiraiSettingSection {
    private val lock = ReentrantLock()

    @Suppress("UNCHECKED_CAST")
    fun <T> getAs(index: Int): T {
        return super.get(index) as T
    }

    fun getInt(index: Int): Int {
        return this.getAs(index)
    }

    fun getDouble(index: Int): Int {
        return this.getAs(index)
    }

    fun getString(index: Int): Int {
        return this.getAs(index)
    }

    fun getFloat(index: Int): Int {
        return this.getAs(index)
    }

    @Synchronized
    override fun saveAsSection(section: Profile.Section) {
        section.clear()
        val integer = AtomicInteger(0)
        this.forEach { a -> section.put(integer.getAndAdd(1).toString(), a) }
    }

    @Throws(IOException::class)
    override fun close() {

    }

}
